### PR TITLE
Implement DBSP health dedupe

### PR DESCRIPTION
**Pull request by leynos**
## Summary
- add retraction handling and duplicate filtering to the DBSP health marshalling path
- expand the health accumulator to separate sequenced and unsequenced events and clamp results
- document the health pipeline and add unit plus behavioural coverage for duplicates, clamping, and healing

## Testing
- make fmt
- make lint
- make test

------
https://chatgpt.com/codex/tasks/task_e_68cdd5cd802483228db4d51e8d28074d

## Summary by Sourcery

Implement end-to-end duplicate detection, retraction, and clamping logic for health in the DBSP synchronization layer. Refactor the circuit-side HealthAccumulator for stable handling of sequenced and unsequenced events, add comprehensive unit and behavioral tests, and update documentation to illustrate the health pipeline.

New Features:
- Add DBSP sync plugin with input/output systems for health state and damage ingestion
- Introduce duplicate filtering for sequenced and unsequenced damage events in the sync layer
- Implement health snapshot retraction handling to keep the DBSP circuit aligned with ECS state
- Track and expose a monotonic duplicate counter in DbspState for telemetry

Enhancements:
- Refactor HealthAccumulator to separate sequenced and unsequenced events, support insert/remove, and clamp health deltas
- Implement Ord and Hash on HealthAccumulator and its archive for stable archive ordering
- Expand DBSP circuit health aggregation to compute net and max sequence values accurately
- Update roadmap and design docs to reflect the completed health and damage pipeline

Documentation:
- Document the DBSP health pipeline and synchronization process with a detailed sequence diagram in design docs

Tests:
- Add unit tests for health delta stream covering idempotency, clamping, max sequence, death flag, and unsequenced accumulation
- Add behavioral tests for the full health pipeline using rspec scenarios covering duplicate filtering, clamping, and healing

### CODE REVIEW

File: src/dbsp_circuit/streams/health.rs [hunk 1]
```diff
@@ -25,37 +25,62 @@ use crate::dbsp_circuit::{DamageEvent, DamageSource, HealthDelta, HealthState};
...
+            .position(|existing| existing == event)
+        {
+            self.unsequenced.swap_remove(pos);
+        }
+        self.has_event = !self.sequenced.is_empty() || !self.unsequenced.is_empty();
+    }
+
     fn merge(&mut self, other: &Self) {
-        for (seq, signed) in &other.entries {
+        for (seq, signed) in &other.sequenced {
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/dbsp_circuit/streams/health.rs:69]: @coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix.

File: tests/health_pipeline.rs [hunk 2]
```diff
@@ -1,54 +1,183 @@
...
+                        seq: None,
+                    };
+                    env.push_damage(damage);
+                    env.update();
+                    assert_eq!(env.current_health(), 50);
+                    assert_eq!(env.duplicate_count(), 4);
+                },
+            );
+        },
+    ));
```
[the remark this answers is not in the document]
Comment (leynos) [reply tests/health_pipeline.rs:182]: @coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix.

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

[tests/health_pipeline.rs](https://github.com/leynos/lille/pull/218/files/d7be86bdb7968ee1386d976aa398f54dd341985b#diff-bdfcbebc3c9da261ed5d486373047a2487a14048ac817ec5cf802984dc244670)
Comment on lines +171 to +190

```
fn new_ticks_consume_damage_once() {
    run_rspec_serial(&rspec::given(
        "duplicate health deltas are idempotent",
        HealthEnv::default(),
        |ctx| {
            ctx.then("new ticks consume damage once", |env| {
                let damage = sequenced_damage_tick_one();
                env.push_damage_twice(damage);
                env.update();
                env.push_damage(damage);
                env.update();
                let next_tick = sequenced_damage_tick_two();
                env.push_damage(next_tick);
                env.update();
                assert_eq!(env.current_health(), 30);
                assert_eq!(env.duplicate_count(), 2);
            });
        },
    ));
}
```

❌ New issue: [Code Duplication](https://codescene.io/projects/68983/delta?repo-id=1157101&review-id=218&biomarker=Code+Duplication&filename=tests%2Fhealth_pipeline.rs&method=)
The module contains 6 functions with similar structure: duplicate_events_within_tick_apply_once,healing_saturates_at_max_health,new_ticks_consume_damage_once,replaying_same_tick_delta_is_ignored and 2 more functions

**Comment by leynos** [comment]
@coderabbitai pause

File: src/dbsp_circuit/tests.rs [hunk 3]
```diff
@@ -127,3 +127,160 @@ fn duplicate_damage_events_idempotent(#[case] seq: Option<u32>) {
...
+        at_tick: 10,
+        seq: Some(4),
+    };
+    let deltas = run_health_delta(health, &[(damage, 1), (heal, 1)]);
+    assert_eq!(deltas.len(), 1);
+    let delta = deltas[0];
+    assert_eq!(delta.delta, -40);
+    assert!(!delta.death);
+    assert_eq!(delta.seq, Some(4));
+}
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/dbsp_circuit/tests.rs:286]: @coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix.

File: src/dbsp_sync/input.rs [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+                Velocity {
+                    entity: id.0,
+                    vx: (v.0 as f64).into(),
+                    vy: (v.1 as f64).into(),
+                    vz: (v.2 as f64).into(),
+                },
+                1,
+            );
+        }
+    }
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/dbsp_sync/input.rs:222]: @coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix.

File: src/dbsp_circuit/tests.rs [hunk 5]
```diff
@@ -127,3 +127,106 @@ fn duplicate_damage_events_idempotent(#[case] seq: Option<u32>) {
...
+        at_tick: 3,
+        seq: None,
+    };
+    let deltas = run_health_delta(health, &[(event, 1)]);
+    assert_eq!(deltas.len(), 1);
+    let delta = deltas[0];
+    assert_eq!(delta.delta, 30);
+    assert!(!delta.death);
+    assert_eq!(delta.seq, None);
+}
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/dbsp_circuit/tests.rs:235]: Has this now been resolved?

File: tests/health_pipeline.rs [hunk 6]
```diff
@@ -1,54 +1,183 @@
...
+                    source: DamageSource::External,
+                    at_tick: 4,
+                    seq: None,
+                };
+                env.push_damage_twice(damage);
+                env.update();
+                assert_eq!(env.current_health(), 50);
+                assert_eq!(env.duplicate_count(), 3);
+            });
+            ctx.then(
```
[the remark this answers is not in the document]
Comment (leynos) [reply tests/health_pipeline.rs:286]: Has this now been resolved?

File: tests/health_pipeline.rs [hunk 7]
```diff
@@ -1,54 +1,183 @@
...
+                    amount: 30,
+                    source: DamageSource::External,
+                    at_tick: 2,
+                    seq: Some(43),
+                };
+                env.push_damage(damage);
+                env.update();
+                assert_eq!(env.current_health(), 30);
+                assert_eq!(env.duplicate_count(), 2);
+            });
```
[the remark this answers is not in the document]
Comment (leynos) [reply tests/health_pipeline.rs:197]: Has this now been resolved?

File: src/dbsp_circuit/tests.rs [hunk 8]
```diff
@@ -120,10 +173,118 @@ fn duplicate_damage_events_idempotent(#[case] seq: Option<u32>) {
...
+        seq: Some(1),
+    };
+    let heal = DamageEvent {
+        entity: 5,
+        amount: 20,
+        source: DamageSource::Script,
+        at_tick: 10,
+        seq: Some(4),
+    };
+    assert_health_delta_test(health, &[(damage, 1), (heal, 1)], -40, false, Some(4));
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/dbsp_circuit/tests.rs:289]: @coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix.
